### PR TITLE
Add output binding wasm

### LINKED ISSUE #2274: New output binding for Wasm
## Describe the feature

We currently have a middleware that allows modifying requests using Wasm.

Since Dapr implements now a Wasm runtime to support that, would it make sense to have the ability to create a binding to allow running code as Wasm?

This would be an output binding with an `invoke` method that allows triggering code from a Wasm file. It would allow turning Dapr into a Wasm runtime easily.

CC: @codefromthecrypt  @daixiang0 @artursouza 

## Release Note

RELEASE NOTE: **ADD** Wasm output binding

### PR DESCRIPTION
**Pull request by codefromthecrypt**
# Description

This adds an output binding for wasm, initially implementing for WASI preview 1. We can later add old WASI (snapshot) or GOOS=js, and in the future new WASI. Also, [soon](https://github.com/golang/go/issues/58141) we can use normal Go to compile as well as TinyGo.

## Issue reference

Closes #2274

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: bindings/wasm/benchmark_test.go
Comment (codefromthecrypt) [inline bindings/wasm/benchmark_test.go:32]: @salaboy will become immortal

**Comment by codefromthecrypt** [comment]
fyi for bench results. Note they are a *lot* worse if you use stdin instead of args, but I didn't add a dimension about that.

```
goos: darwin
goarch: arm64
pkg: github.com/dapr/components-contrib/bindings/wasm
BenchmarkExample
BenchmarkExample-12    	  101841	     14366 ns/op
PASS
```

File: bindings/wasm/output.go
Comment (ItalyPaleAle) [inline bindings/wasm/output.go:98]: Is `_ = out.runtime.Close(ctx)` not needed here?

File: bindings/wasm/output.go
Comment (ItalyPaleAle) [inline bindings/wasm/output.go:152]: Rather than encoding to JSON and back, please use our `DecodeMetadata` function (example: https://github.com/dapr/components-contrib/blob/76e33137d95f60f182a21505d361ba633afd44ed/state/sqlite/sqlite_metadata.go#L57) which has a bunch of advantages

File: bindings/wasm/output.go
Comment (ItalyPaleAle) [inline bindings/wasm/output.go:120]: If `req.Metadata["args"]` is not defined, you don't want to pass `out.module.Name()` as args?

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
This is awesome, just some small tweaks please :)

File: bindings/wasm/output.go
Comment (yaron2) [inline bindings/wasm/output.go:139]: Is this really a `Get` operation? we can support any operation name that matches what is being done as closely as possible - perhaps `Execute`?

**Comment by codefromthecrypt** [comment]
Thanks for all the awesome feedback. I will work it in.

I have a question about an important missing detail, which are ENV variables. Most of the time, the only ENV you need are at init scope. e.g. setting TMPDIR to something besides /tmp (though we haven't plugged any persistence in, yet). Some things use them at invoke time, but yeah using them at request time can also be problematic as people can override your init-level settings (unless we forbid that)

So, two questions

1. how should we allow env variables to be passed through this? (I think in one case we'd have to serialize as a string)?
2. If so, should we allow invoke to override the same env key set during init?

I think for first commit we can stick with args/env and later I can help with filesystem. For example, if we want to allow output bindings to mount / to a directory on the host, and if so do we want to magically handle /tmp for invocations so each has their own.. basically, this discussion is big enough to be another PR, but I wanted to start the brain nag.

File: bindings/wasm/output.go
Comment (codefromthecrypt) [reply bindings/wasm/output.go:88]: so this context was made to allow trace context propagation. e.g. know which request closed your runtime, so I think we'll document that close will operate even if the context is canceled. OTOH it is fine do do what you are saying as we haven't actually exposed a trace interface, yet.

File: bindings/wasm/output.go
Comment (codefromthecrypt) [reply bindings/wasm/output.go:116]: I'll add tests. no, an empty isn't desired indeed!

File: bindings/wasm/output.go
Comment (codefromthecrypt) [reply bindings/wasm/output.go:120]: so arg0 is the program name.. it is a bit arbitrary what we choose here, but some things expect that even if there are no args... wdyt?

File: bindings/wasm/output.go
Comment (codefromthecrypt) [reply bindings/wasm/output.go:98]: good eye.

**Comment by codefromthecrypt** [comment]
another thing I need feedback on. context cancelation/timeout integration has a high overhead, some factor of ten percent. Most runtimes don't enable things like this by default due to the performance, but it is certainly safer to do that.

Should we enable context cancelation/timeout integration by default in dapr? or should we make this a config?

Either way, we can with a config, and see results even though our demo wasm is a toy. It would be nice to see some representative WASI wasm someone wants to run when this feature is out.

File: bindings/wasm/output.go
Comment (ItalyPaleAle) [reply bindings/wasm/output.go:120]: I don't have a problem with arg0.

What I was trying to say is that you set `moduleConfig = moduleConfig.WithArgs(argsSlice...)` only within the block with `if req.Metadata["args"]`. If `req.Metadata["args"]` is empty, you don't want to set `moduleConfig.WithArgs`?

**Comment by ItalyPaleAle** [comment]
@codefromthecrypt 

With regards to filesystem access and env vars, I think we can postpone that for later. We ought to be careful with exposing env vars and allowing unfettered access to the filesystem, because that could allow accessing secrets that may not be desirable. Do you think that env var and fs access are a must?

As for context cancellation, are you talking about cancelling Invoke() via cancelling the context? Ideally, we'd like to have that supported. Maybe it could be an opt-out thing, which is enabled by default but you can pass a metadata option (in the request) to not enable that?

File: bindings/wasm/output.go
Comment (codefromthecrypt) [reply bindings/wasm/output.go]: made it more explicit, and tested it. please re-open this if the direction isn't correct.

**Comment by codefromthecrypt** [comment]
PTAL: we're ready except how to handle ENV unless we want to keep it out of this PR.

File: bindings/wasm/benchmark_test.go
Comment (ItalyPaleAle) [inline bindings/wasm/benchmark_test.go]: No more @salaboy? :)

**Comment by codefromthecrypt** [comment]
whoops I forgot to integrate context cancelation. Lemme do that now.

**Comment by codefromthecrypt** [comment]
@ItalyPaleAle I left this enabled and we can later decide to add/name a property to disable context cancelation. on my laptop the dummy test wasm takes 14us with cancelation enabled and 12us without.

Lemme know what we need to do next on this PR and thanks for the lightning feedback!

File: bindings/wasm/benchmark_test.go
Comment (codefromthecrypt) [reply bindings/wasm/benchmark_test.go]: :) I decided to use the benchmark to test baseline overhead, like the minimal for any wasi function. salaboy still lives forever in the test cases :D

**Comment by ItalyPaleAle** [comment]
I fixed the linter issues, and now LGTM! Thanks for the changes

However, I'd like to get @berndverst's approval on this PR too before we merge it :)

File: bindings/wasm/benchmark_test.go
Comment (salaboy) [reply bindings/wasm/benchmark_test.go]: 😂

File: bindings/wasm/output.go
Comment (pkedy) [inline bindings/wasm/output.go:81]: Nit: duplicate `WithSysNanosleep()`. No negative affect though.

File: bindings/wasm/output.go
Comment (pkedy) [inline bindings/wasm/output.go:142]: @codefromthecrypt You know best. Even though `_start` runs to completion, do you need to call `module.Close()` to clean up any resources in wazero?

Would it make sense to add `defer module.Close()`?

**Review by pkedy — DISMISSED** [review]
A few comments (mainly curious about the need for `module.Close()`) but LGTM.

File: bindings/wasm/output.go
Comment (berndverst) [inline bindings/wasm/output.go:79]: Remove this duplicate option method

```suggestion
```

**Review by berndverst — APPROVED** [review]

File: bindings/wasm/output.go
Comment (codefromthecrypt) [reply bindings/wasm/output.go:141]: good eye!

Comment (codefromthecrypt) [reply bindings/wasm/output.go:141]: https://github.com/dapr/components-contrib/pull/2697

**Pull request merged**
